airflow/io/__init__.py

```diff
@@ -0,0 +1,297 @@
+        self._location = location
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the total length of the file, in bytes."""
+
+    @property
+    def location(self) -> str:
+        """The fully-qualified location of the input file."""
+        return self._location
+
+    @abstractmethod
+    def exists(self) -> bool:
+        """Check whether the location exists.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+        """
+
+    @abstractmethod
+    def open(self, seekable: bool = True) -> InputStream:
+        """Return an object that matches the InputStream protocol.
+
+        Args:
+            seekable: If the stream should support seek, or if it is consumed sequential.
+
+        Returns:
+            InputStream: An object that matches the InputStream protocol.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+            FileNotFoundError: If the file at self.location does not exist.
+        """
+
+
+class OutputFile(ABC):
+    """A base class for OutputFile implementations.
+
+    Args:
+        location (str): A URI or a path to a local file.
+
+    Attributes:
+        location (str): The URI or path to a local file for an OutputFile instance.
+        exists (bool): Whether the file exists or not.
+    """
+
+    def __init__(self, location: str):
+        self._location = location
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the total length of the file, in bytes."""
+
+    @property
+    def location(self) -> str:
+        """The fully-qualified location of the output file."""
+        return self._location
+
+    @abstractmethod
+    def exists(self) -> bool:
+        """Check whether the location exists.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+        """
+
+    @abstractmethod
+    def to_input_file(self) -> InputFile:
+        """Return an InputFile for the location of this output file."""
+
+    @abstractmethod
+    def create(self, overwrite: bool = False) -> OutputStream:
+        """Return an object that matches the OutputStream protocol.
+
+        Args:
+            overwrite (bool): If the file already exists at `self.location`
+                and `overwrite` is False a FileExistsError should be raised.
+
+        Returns:
+            OutputStream: An object that matches the OutputStream protocol.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+            FileExistsError: If the file at self.location already exists and `overwrite=False`.
+        """
+
+
+class FileIO(ABC):
+    """A base class for FileIO implementations."""
+
+    @abstractmethod
+    def new_input(self, location: str, conn_id: str | None) -> InputFile:
```

Comment:
Coding question... does it make sense to instantiate the `FileIO` class just to open `new_input()`? Shall this be rather a static method or a module level entry point function?

Reply:
This part is a straight port from `pyiceberg` and I am wondering the same. It seems more intuitive to use `FsspecFileIO.new_input` or even `FileIO.new_input`